lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -3579,6 +3341,97 @@ LogicalResult FIRRTLLowering::visitStmt(SkipOp op) {
   return success();
 }
 
+void FIRRTLLowering::lowerRegConnect(Operation *regOp, Value dest,
+                                     Value srcVal) {
+  // Unroll all accesses since they must be traversed in inverse order.
+  SmallVector<Operation *> accesses;
+  while (dest) {
+    TypeSwitch<Operation *, void>(dest.getDefiningOp())
+        .Case<SubfieldOp, SubindexOp>([&](auto op) {
+          accesses.push_back(op);
+          dest = op.getInput();
+        })
+        .Default([&](auto op) {
+          assert(op == regOp && "invalid lowering");
+          dest = {};
+        });
+  }
+
+  // Construct the updated value.  For each access, starting with the
+  // outermost ones, the new value is computed by updating one of its fields.
+  // The value of the field itself is determined by updating the previous
+  // value according to the subsequent access operation.
+  std::function<Value(Value, unsigned)> inject = [&](Value base, unsigned i) {
+    if (auto op = dyn_cast<SubfieldOp>(accesses[i])) {
+      auto ty = hw::type_cast<hw::StructType>(base.getType());
+      auto field = ty.getElements()[op.getFieldIndex()];
+      auto name = field.name;
+
+      // Determine the value of the new field.
+      Value value;
+      if (i == 0) {
+        value = srcVal;
+      } else {
+        auto oldField =
+            builder.create<hw::StructExtractOp>(field.type, base, field.name);
+        value = inject(oldField, i - 1);
+      }
+
+      return builder.create<hw::StructInjectOp>(ty, base, name, value)
+          .getResult();
+    }
+    if (auto op = dyn_cast<SubindexOp>(accesses[i])) {
+      auto ty = hw::type_cast<hw::ArrayType>(base.getType());
+      auto elemTy = ty.getElementType();
+
+      auto size = ty.getSize();
+      auto index = op.getIndex();
+      auto indexWidth = std::max(1u, llvm::Log2_64_Ceil(size));
```

Comment:
Mega-nit: this looks safe to me as `1u` is 32-bit on both Linux and Windows. However, it seems slightly dirty to mix `1u` with the `uint64_t` returned by `Log2_64_Ceil`. 

Also, this will return `64` if size == 0. Is that intended?

Reply:
Thanks for spotting this. `Log2_64_Ceil` returns unsigned, so `1u` is required for `std::max` to type-check. I switched over to `getBitWidthFromVectorSize`, which is used in other places on vector types without guarding for 0. We should address this problem in another PR.